src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2376,44 +2426,175 @@ private void SelectVoice(string voice)
             NavigateToMenu();
         }
 
-        private void StepList(string command)
+        public async Task StepList(KeyValue singleKeyValue, List<string> multiKeySelection)
         {
-            if (!string.IsNullOrEmpty(command))
+            if (!string.IsNullOrEmpty(singleKeyValue.String))
             {
-                KeyValue singleKeyValue;
-                string[] stringSeparators = new string[] { "<" };
-                foreach (var vStep in command.Split(stringSeparators, StringSplitOptions.None).ToList())
+                Log.InfoFormat("StepList called with singleKeyValue [{0}]", singleKeyValue.String);
+
+                List<string> commandList = singleKeyValue.String.Split('<').ToList();
+                if (commandList.Any())
+                {
+                    await StepListCommands(singleKeyValue, multiKeySelection, commandList, 0);
+                }
+            }
+        }
+
+        private async Task StepListCommands(KeyValue singleKeyValue, List<string> multiKeySelection, List<string> commandList, int nestLevel)
+        {
+            Log.InfoFormat("StepListCommands called with a list of {0} commands", commandList.Count);
+            var keyDownList = new List<string>();
+
+            if (nestLevel == 0)
+            {
+                keyStateService.KeyRunningStates[singleKeyValue].Value = true;
+                keyStateService.KeyDownStates[singleKeyValue].Value = KeyDownStates.Down;
+
+                keyDownList.AddRange(commandList.FindAll(x => x.StartsWith("KeyToggle>")));// || x.StartsWith("KeyDown>")););
+            }
+            while (commandList.Any())
+            {
+                //if a key stop has been triggerred, then stop processing and return
+                if (!keyStateService.KeyRunningStates[singleKeyValue].Value)
+                {
+                    //if at the top level execute KeyUp processing
+                    if (nestLevel == 0 && keyDownList != null && keyDownList.Any())
+                    {
+                        foreach (var downKey in keyDownList)
+                        {
+                            var keyUpCandidate = new KeyValue(downKey.Substring(downKey.IndexOf(">")));
+                            if (keyStateService.KeyDownStates[keyUpCandidate].Value != KeyDownStates.Up)
+                                keyStateService.KeyDownStates[keyUpCandidate].Value = KeyDownStates.Up;
+                        }
+                    }
+                    return;
+                }

```

Comment:
The start of new logic to handle long running keys and resetting of key states. It's not done yet.